PatchBuilder

TimurSirkin opened the pull request.

TimurSirkin committed:
Add PatchBuilder files and Templates.ini

Nikolay2k commented on include/PatchBuilder/PatchBuilder.h:
```diff
@@ -1,10 +1,62 @@
-  PatchBuilder();
-  ~PatchBuilder();
+  PatchBuilder() = default;
+  PatchBuilder(const string pPatchListFullName, DBProvider &pProvider, const string pTemplatesFullName = ""); // Constructor
```
Все параметры(кроме int) необходимо передавать по ссылке, т.е. через &. 
`PatchBuilder(const string &pPatchListFullName, DBProvider &pProvider, const string &pTemplatesFullName = "")`
В этом случае параметры не будут копироваться в стек. Правка самого тела метода не требуется. Прошу исправить это и в других методах.

- [x] параметры по ссылке через &

Nikolay2k commented on include/PatchBuilder/PatchBuilder.h:
```diff
@@ -1,10 +1,62 @@
+	string patchListFullName; // Directory and name of PatchList
+	string templateString = ""; // Teplate text
+	string logFileFullName; // Directory and name of file for logs
+	DBProvider *provider; // Ptr to provider class
```
Для чего используется этот указатель? Создавайте объект сразу. Либо используйте "умный" указатель `std::shared_ptr` или `std::unique_ptr`

- [x] Создавайте объект сразу

TimurSirkin replied:
Не очень понимаю зачем.
Мне DBProvider приходит извне уже созданный.

Nikolay2k replied:
Это не правильно. Он не должен ниоткуда приходить. DBProvider должен быть членом класса

TimurSirkin replied:
Как скажете - переделаю

Nikolay2k commented on include/PatchBuilder/PatchBuilder.h:
```diff
@@ -1,10 +1,62 @@
+  void addLog(const string message) const; // Add message in log file
+
+private:
+	string patchListFullName; // Directory and name of PatchList
```
Просьба использовать snake_style для имен переменных и параметров

- [x] patchListFullName -> patch_list_full_name

Nikolay2k commented on include/PatchBuilder/PatchBuilder.h:
```diff
@@ -1,10 +1,62 @@
+	DBProvider *provider; // Ptr to provider class
+	bool isWithErrors = false; // Become true after some error
+	bool isWithWarnings = false; // Become true after some warning
+	scriptDataVectorType getScriptDataVector(const objectDataVectorType &objectDataVector) const;// Getting vector of scripts, created by DBProvider
```
Просьба использовать snake_style для имен переменных и параметров

- [x] objectDataVector -> object_data_vector

Nikolay2k commented on src/PatchBuilder.cpp:
```diff
@@ -1,5 +1,454 @@
+void PatchBuilder::createInstallPocket(const string directory, const scriptDataVectorType &scriptDataVector) const
+{
+	ofstream outputInstallScriptBat(directory + "/" + INSTALL_SCRIPT_NAME_BAT);
+	ofstream outputInstallScriptSh(directory + "/" + INSTALL_SCRIPT_NAME_SH);
```
Постарайтесь использовать более короткие понятные имена

- [x] `outputInstallScriptBat` -> `install_bat`

- [x] `outputInstallScriptSh` -> `install_sh`

Nikolay2k commented on src/PatchBuilder.cpp:
```diff
@@ -1,5 +1,454 @@
+
+		// Creating install command
+		string installString = string("psql -a -U ") + "%1" + " -d " + "%2" + " -h " + "%3" + " -p " + "%4" " -f " + data.schema + "/";
+		if (data.type != "")
```
для проверки пустая ли строка используйте метод `empty()`

- [x] `empty()`

Nikolay2k commented on src/PatchBuilder.cpp:
```diff
@@ -1,5 +1,454 @@
+		}
+		output << endl;
+	}
+}
```
Необходимо закрыть файл

- [x] `output.close()`

TimurSirkin committed:
Refactoring

TimurSirkin committed:
Add empty() method in condition

Nikolay2k commented on src/PatchBuilder_exe.cpp:
```diff
@@ -1,4 +1,145 @@
+		{
+			if (currentFlag == flag)
+			{
+				return &arg; // Return the argument with the right flag
```
Возвращайте не ссылку на элемент, а сам элемент.
 
- [x] `return arg;`

TimurSirkin replied:
Сам метод должен возвращать адрес, а не элемент.
Сделал это для того, чтобы можно было возвращать nullptr, если элемент не найден.
Могу, конечно, ввести булевскую переменную, которую буду тоже возвращать.

Nikolay2k commented on src/PatchBuilder_exe.cpp:
```diff
@@ -1,4 +1,145 @@
+	PatchBuilder builder{};
+	try
+	{
+		DBProvider provider(getArgumentByKey(args, KEY_CONNECTION)->value);
```
`DBProvider provider` должен быть приватным членом класса `PatchBuilder`. При создании объекта `builder` необходимо передавать не сам провайдер, а только строку коннекта. А создание объекта `DBProvider` скрыть внутри класса `PatchBuilder`. Тогда это позволит обращаться к методам `DBProvider` не через указатель.

- [x] скрыть использование `DBProvider` внутри `PatchBuilder`

Nikolay2k commented on src/PatchBuilder.cpp:
```diff
@@ -1,5 +1,454 @@
+	}
+	const string message =  "Script vector created\n";
+	cout << message;
+	addLog(message);
```
У вас везде встречается похожие три строки. Это напрашивается на создания отдельной функции для этого.

- [x] `show_message`

Nikolay2k commented on include/PatchBuilder/PatchBuilder.h:
```diff
@@ -1,10 +1,62 @@
+	bool isContains(const ObjectData data, const string &scriptText); // Returns true, if the object was found in the script
+	objectDataVectorType getPatchListVector() const; // Getting vector of objects that contains a patch
+	void createObjectList(const scriptDataVectorType &objectDataVector, const string directory) const; // Creating of ObjectList
+	static void remove(objectDataVectorType &objectDataVector_first, const objectDataVectorType &objectDataVector_second); // Removing elements of second vector from first vector
```
Очень длинные строки. Включите стиль форматирования LLVM. В Visual Studio для этого надо Средство->Параметры->Текстовый редактор->C/C++->Форматирование->Общие->Стиль форматирования->LLVM
После этого в каждом файле проведите форматирование. Нажать сначала Ctrl+K затем Ctrl+D

- [x] LLVM

TimurSirkin committed:
Changes required from pull request

Applied LLVM code style
Change return values of getArgByValue and getArgByKey methods
Rename some variables and parameters
Add showMessage method
DBProvider now created inside PatchBuilder by connection arguments from constructor
